netket/experimental/driver/tdvp_schmitt.py

```diff
@@ -225,8 +225,9 @@ def _impl(parameters, n_samples, E_loc, S, rhs_coeff, rcond, rcond_smooth, snr_a
     QEdata = Q.conj() * ΔE_loc
     rho = V.conj().T @ F
 
-    # Compute the SNR according to Eq. 21
-    snr = jnp.abs(rho) * jnp.sqrt(n_samples) / jnp.sqrt(stats.var(QEdata, axis=0))
+    # Compute the SNR according to Eq. 21 but taking care of where sigma_k is zero
+    sigma_k = jnp.maximum(jnp.sqrt(stats.var(QEdata, axis=0)), rcond)
+    snr = jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k
```

Comment:
I'm not against this approach, but the current documentation states that `rcond` is the (relative) cutoff for small eigenvalues. $$ \sigma_k $$ here is instead something like the variance/uncertainty of one eigenvalue, which might very well be large or small...
while both are cutoffs, is it ok to use the same number in both places? I have some doubts... 

--
My reasoning is the following:

In general, if $$\sigma_k$$ is very small, the SNR is well defined and will be extremely large, meaning that in `regularizer2`, below, we will not be regularising it at all.

I think the problem here is really only for $$\sigma_k=0$$ and $$\rho = 0 $$ , because in that case we get 0/0 which is NaN.  I'm not sure what other situation might yield a NaN in the current code.
If only $$\sigma_k$$ is small, then the snr will be something divided by 0 which is infinity, which is correctly handled in the code later. 

Indeed, if an eigenvalue is very small (0) and the variance is very small (0) it's impossible to define what is the SNR (there is no signal, but no noise either).
What should we be doing in that case? As there is no signal, we should not be moving.

To recap: I think the current code is correct for even 0 snr, and is only 'broken' when snr=0 and $$\rho=0$$.
In this case, I believe the right thing to do would be to set `eta_p=0` (not move) explicitly...

What do you think?

Reply:
Ok, I completely agree. Do we also agree that this taken care by
```python3
    sigma_k_is_zero = jnp.isclose(sigma_k, 0, atol=1e-15)
    rho_is_zero = jnp.isclose(rho, 0, atol=1e-15)
    snr = jnp.where(
        sigma_k_is_zero & rho_is_zero,
        0.0,
        jnp.where(
            sigma_k_is_zero,
            jnp.inf,
            jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k,
        )
    )

```
? I just want to avoid the code carries out the NaNs just to filter them only at `eta_p`. This way, `eta_p` will be zero due to the `rho` multiplication at `eta_p = ev_inv * regularizer2 * rhs_coeff * rho, and the NaNs are avoided from the start.

Reply:
Is there a particular reason you want to avoid nan being generated?

Regardless, in the code you wrote above you could drop the `jnp.where` in
```python
        jnp.where(
            sigma_k_is_zero,
            jnp.inf,
            jnp.abs(rho) * jnp.sqrt(n_samples) / sigma_k,
        )
```

because you'd get infinity/very large number anyway?

```python
In [17]: sigma_k=jnp.array([0.0, 1e-16])
In [18]: sigma_k_is_zero = jnp.isclose(sigma_k, 0, atol=1e-15)
In [19]: 1/sigma_k
Out[19]: Array([   inf, 1.e+16], dtype=float32)
```
what are you trying to avoid?